File: frontend/src/types/manifest.ts
import type { StemRole, StemStatus } from "@/api/stems";

export interface SongManifest {
  song: {
    id: number;
    title: string;
    artist: string;
    slug: string;
    originalKey: number;
    durationMs: number | null;
    sampleRate: number | null;
  };
  keyVariants: SongKeyVariant[];
  selectedKeyId: number | null;
  playable: boolean;
  stems: StemManifestItem[];
  playerSettings?: PlayerSettings;
}

export interface SongKeyVariant {
  id: number;
  semitoneOffset: number;
  isOriginal: boolean;
  status: "draft" | "ready" | "error";
  playable: boolean;
  errorMessage: string | null;
}

export interface PlayerSettings {
  stemGainDefaultDb: number;
  stemGainMinDb: number;
  stemGainMaxDb: number;
  stemGainStepDb: number;
  focusGainDefaultDb: number;
  focusGainMinDb: number;
  focusGainMaxDb: number;
  backgroundGainDefaultDb: number;
  backgroundGainMinDb: number;
  backgroundGainMaxDb: number;
}

export interface StemManifestItem {
  id: number;
  name: string;
  role: StemRole;
  key: number | null;
  focusable: boolean;
  status: StemStatus;
  url: string | null;
  codec: string | null;
  container: "m4a" | null;
  channels: number | null;
  sampleRate: number | null;
  durationMs: number | null;
  fileSizeBytes: number | null;
  bitrateKbps: number | null;
  errorMessage: string | null;
}

export type PlayableStemManifestItem = StemManifestItem & {
  status: "ready";
  url: string;
};

export function isPlayableStem(stem: StemManifestItem): stem is PlayableStemManifestItem {
  return stem.status === "ready" && stem.url !== null;
}

File: backend/app/schemas/manifest.py
from pydantic import BaseModel, ConfigDict, Field

from app.domain import StemRole, StemStatus
from app.schemas.settings import PlayerSettings


class ManifestSong(BaseModel):
    id: int
    title: str
    slug: str
    duration_ms: int | None = Field(default=None, alias="durationMs")
    sample_rate: int | None = Field(default=None, alias="sampleRate")

    model_config = ConfigDict(populate_by_name=True)


class ManifestStem(BaseModel):
    id: int
    name: str
    role: StemRole
    status: StemStatus
    url: str | None = None
    codec: str | None = None
    container: str | None = None
    channels: int | None = None
    sample_rate: int | None = Field(default=None, alias="sampleRate")
    duration_ms: int | None = Field(default=None, alias="durationMs")
    file_size_bytes: int | None = Field(default=None, alias="fileSizeBytes")
    bitrate_kbps: int | None = Field(default=None, alias="bitrateKbps")
    error_message: str | None = Field(default=None, alias="errorMessage")

    model_config = ConfigDict(populate_by_name=True)


class SongManifest(BaseModel):
    song: ManifestSong
    playable: bool
    stems: list[ManifestStem]
    player_settings: PlayerSettings = Field(alias="playerSettings")

    model_config = ConfigDict(populate_by_name=True)
